
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="installation.html">
      
      
        <link rel="next" href="hybrid.html">
      
      
      <link rel="icon" href="spades.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Basic input - SPAdes Assembly Toolkit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#spades-input" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="SPAdes Assembly Toolkit" class="md-header__button md-logo" aria-label="SPAdes Assembly Toolkit" data-md-component="logo">
      
  <img src="spades.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SPAdes Assembly Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Basic input
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="SPAdes Assembly Toolkit" class="md-nav__button md-logo" aria-label="SPAdes Assembly Toolkit" data-md-component="logo">
      
  <img src="spades.png" alt="logo">

    </a>
    SPAdes Assembly Toolkit
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="getting-started.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quick start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="installation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    SPAdes input data
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            SPAdes input data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Basic input
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="input.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Basic input
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#paired-read-libraries" class="md-nav__link">
    <span class="md-ellipsis">
      Paired read libraries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#unpaired-single-read-libraries" class="md-nav__link">
    <span class="md-ellipsis">
      Unpaired (single-read) libraries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reading-ncbi-sra-files" class="md-nav__link">
    <span class="md-ellipsis">
      Reading NCBI SRA files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="hybrid.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hybrid assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="datatypes.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tips on the parameters
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="running.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command line options
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="output.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SPAdes output
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="hmm.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    HMM-guided mode
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="rna.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Transcriptome assembly
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="hpc.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running SPAdes across multiple nodes of a cluster
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="binspreader.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Binning refining
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="pathracer.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    HMM mapping on assembly graph
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="spaligner.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sequence to graph alignment
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="standalone.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SPAdes tools
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="citation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="feedback.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#paired-read-libraries" class="md-nav__link">
    <span class="md-ellipsis">
      Paired read libraries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#unpaired-single-read-libraries" class="md-nav__link">
    <span class="md-ellipsis">
      Unpaired (single-read) libraries
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reading-ncbi-sra-files" class="md-nav__link">
    <span class="md-ellipsis">
      Reading NCBI SRA files
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="spades-input">SPAdes input</h1>
<p>SPAdes takes as input paired-end reads, mate-pairs and single (unpaired) reads in FASTA and FASTQ (can be gzipped) formats. Additionally, SPAdes can read <a href="https://www.ncbi.nlm.nih.gov/sra/docs/sra-data-formats/">NCBI SRA</a> files directly. If you're compiling SPAdes from sources, make sure to read <a href="input.html#reading-ncbi-sra-files">corresponding section</a> for more information.</p>
<p>For IonTorrent data SPAdes also supports unpaired reads in unmapped BAM format (like the one produced by Torrent Server).</p>
<p>However, in order to run read error correction, reads should be in FASTQ or BAM format.
Sanger, Oxford Nanopore and PacBio CLR reads can be provided in any formats since SPAdes does not run error correction for these types of data.</p>
<p>To run SPAdes you need at least one library of the following types:</p>
<ul>
<li>Illumina paired-end/high-quality mate-pairs/unpaired reads</li>
<li>IonTorrent paired-end/high-quality mate-pairs/unpaired reads</li>
<li>PacBio CCS reads</li>
</ul>
<p>Illumina and IonTorrent libraries should not be assembled together. All other types of input data are compatible.
SPAdes should not be used if only PacBio CLR, Oxford Nanopore, Sanger reads or additional contigs are available.</p>
<p>SPAdes supports mate-pair only assembly. However, we recommend to use only high-quality mate-pair libraries in this case
(e.g. that do not have a paired-end part). We tested the mate-pair-only pipeline using Illumina Nextera mate-pairs. See more <a href="running.html#specifying-multiple-libraries">here</a>.</p>
<p>Notes:</p>
<ul>
<li>It is strongly recommended to provide multiple paired-end and mate-pair libraries according to their insert size (from smallest to longest).</li>
<li>It is not recommended to run SPAdes on PacBio reads with low coverage (less than 5).</li>
<li>We suggest not to run SPAdes on PacBio reads for large genomes.</li>
</ul>
<h2 id="paired-read-libraries">Paired read libraries</h2>
<p>By using command line interface, you can specify (1) paired-end, (2) standard mate-pair and (3) high-quality mate-pair libraries.
You can provide up to nine libraries of each type. Libraries can be used in any combination, but recommend not to assemble low-quality mate-pairs alone.
If you wish to use more libraries, you can use <a href="running.html#specifying-multiple-libraries-with-yaml-data-set-file">YAML data set file</a>.</p>
<p>By default, SPAdes assumes that paired-end and high-quality mate-pair reads have forward-reverse (fr) orientation and usual mate-pairs have reverse-forward
(rf) orientation. However, different orientations can be indicated for any library by using SPAdes options.</p>
<p>To distinguish reads in pairs we refer to them as left and right reads. For forward-reverse orientation, the forward reads correspond to the left reads and the reverse reads, to the right. Similarly, in reverse-forward orientation left and right reads correspond to reverse and forward reads, respectively, etc.</p>
<p>Each paired-end or mate-pair library can be stored in several files or several pairs of files. Paired reads can be organized in two different ways:</p>
<ul>
<li>In file pairs. In this case left and right reads are placed in different files and must go in the same order.
I.e. for every left read at line X in the first file the corresponding right read from the pair must be at line X in the second file.</li>
<li>In interleaved files. In this case, the reads are interlaced, so that each right read goes after the corresponding paired left read.</li>
</ul>
<p>For example, Illumina produces paired-end reads in two files: <code>R1.fastq</code> and <code>R2.fastq</code>. If you choose to store reads in file pairs make sure that for every read from <code>R1.fastq</code> the corresponding paired read from <code>R2.fastq</code> is placed in the respective paired file on the same line number. If you choose to use interleaved files, every read from <code>R1.fastq</code> should be followed by the corresponding paired read from <code>R2.fastq</code>.</p>
<p>If adapter and/or quality trimming software has been used prior to assembly, files with the orphan reads can be provided as "single read files" for the corresponding read-pair library.</p>
<p>If you have merged some of the reads from your paired-end (not mate-pair or high-quality mate-pair) library (using tools s.a. <a href="https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmerge-guide/">BBMerge</a> or <a href="https://bitbucket.org/yaoornl/align_test/overview">STORM</a>), you should provide the file with resulting reads as a "merged read file" for the corresponding library.
Note that non-empty files with the remaining unmerged left/right reads (separate or interlaced) <strong>must</strong> be provided for the same library (for SPAdes to correctly detect the original read length).</p>
<p>In an unlikely case some of the reads from your mate-pair (or high-quality mate-pair) library are "merged", you should provide the resulting reads as a SEPARATE single-read library.</p>
<p>See <a href="running.html#examples">examples</a>.</p>
<h2 id="unpaired-single-read-libraries">Unpaired (single-read) libraries</h2>
<p>By using the command line interface, you can specify up to nine different single-read libraries. To input more libraries, you can use <a href="running.html#specifying-multiple-libraries-with-yaml-data-set-file">YAML data set file</a>.</p>
<p>Single libraries are assumed to have high quality and reasonable coverage. For example, you can provide PacBio CCS reads as a single-read library.</p>
<p>Note, that you should not specify PacBio CLR, Sanger reads or additional contigs as single-read libraries, each of them has a separate <a href="running.html#input-data">option</a>.</p>
<p>See <a href="running.html#examples">examples</a>.</p>
<h2 id="reading-ncbi-sra-files">Reading NCBI SRA files</h2>
<p><a href="https://www.ncbi.nlm.nih.gov/sra/docs/sra-data-formats">NCBI SRA</a> uses a binary format to store reads. SPAdes has the ability to read these files directly, so there is no need to convert them into intermediate FASTQ files. However, this brings additional dependency on NCBI SDK, which may cause compatibility issues. Overall, to assemble reads from SRA files make sure that:</p>
<ul>
<li>SPAdes is compiled with NCBI SDK support enabled (pass <code>-DSPADES_USE_NCBISDK=ON</code> to <code>spades_compile.sh</code>, disabled by default). No action is required if you use pre-compiled release binaries.</li>
<li>Input files have <code>.sra</code> extension.</li>
<li>Input files are provided as interlaced (in case of paired-end reads) or as single-end library.</li>
<li>Read error correction does not support reading SRA files, <code>--only-assembler</code> command line option is mandatory.</li>
</ul>
<p>SRA files can contain reads that were aligned to a reference genome. Reading such files without a reference can be slow and SPAdes will produce a warning in this case. We recommend to convert such files to FASTQ first.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": ".", "features": ["content.code.copy"], "search": "assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>