
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="binspreader.html">
      
      
        <link rel="next" href="spaligner.html">
      
      
      <link rel="icon" href="spades.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>HMM mapping on assembly graph - SPAdes Assembly Toolkit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#pathracer-racing-profile-hmm-paths-on-assembly-graph" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="SPAdes Assembly Toolkit" class="md-header__button md-logo" aria-label="SPAdes Assembly Toolkit" data-md-component="logo">
      
  <img src="spades.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SPAdes Assembly Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              HMM mapping on assembly graph
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="SPAdes Assembly Toolkit" class="md-nav__button md-logo" aria-label="SPAdes Assembly Toolkit" data-md-component="logo">
      
  <img src="spades.png" alt="logo">

    </a>
    SPAdes Assembly Toolkit
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="getting-started.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quick start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="installation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    SPAdes input data
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            SPAdes input data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="input.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basic input
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="hybrid.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hybrid assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="datatypes.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tips on the parameters
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="running.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command line options
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="output.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SPAdes output
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="hmm.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    HMM-guided mode
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="rna.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Transcriptome assembly
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="hpc.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running SPAdes across multiple nodes of a cluster
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="binspreader.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Binning refining
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    HMM mapping on assembly graph
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="pathracer.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    HMM mapping on assembly graph
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compilation" class="md-nav__link">
    <span class="md-ellipsis">
      Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pathracer-tool" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer tool
    </span>
  </a>
  
    <nav class="md-nav" aria-label="pathracer tool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pathracer-command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer command line options
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pathracer-output" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pathracer-seq-fs-tool" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer-seq-fs tool
    </span>
  </a>
  
    <nav class="md-nav" aria-label="pathracer-seq-fs tool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pathracer-seq-fs-command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer-seq-fs command line options
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pathracer-seq-fs-output" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer-seq-fs output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-files-format" class="md-nav__link">
    <span class="md-ellipsis">
      Output files format
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    <span class="md-ellipsis">
      Examples
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="spaligner.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sequence to graph alignment
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="standalone.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SPAdes tools
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="citation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="feedback.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compilation" class="md-nav__link">
    <span class="md-ellipsis">
      Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pathracer-tool" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer tool
    </span>
  </a>
  
    <nav class="md-nav" aria-label="pathracer tool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pathracer-command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer command line options
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pathracer-output" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pathracer-seq-fs-tool" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer-seq-fs tool
    </span>
  </a>
  
    <nav class="md-nav" aria-label="pathracer-seq-fs tool">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pathracer-seq-fs-command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer-seq-fs command line options
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pathracer-seq-fs-output" class="md-nav__link">
    <span class="md-ellipsis">
      pathracer-seq-fs output
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-files-format" class="md-nav__link">
    <span class="md-ellipsis">
      Output files format
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    <span class="md-ellipsis">
      Examples
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="pathracer-racing-profile-hmm-paths-on-assembly-graph">PathRacer: racing profile HMM paths on assembly graph</h1>
<h2 id="overview">Overview</h2>
<p>PathRacer is assembly graph against profile HMM aligning tool supporting
both <em>local-local</em> and <em>global-local</em> (aka <em>glocal</em>) alignment and both nucleotide and amino acid profile HMMs.
The tool finds all proper alignments rather than only the best one.
That allows extracting all genes satisfying HMM gene model from the assembly.</p>
<p><strong>PathRacer</strong> is a tool for alignment of assembly graph against pHMM. It provides
the set of <em>k</em> most probable paths traversed by a HMM through the whole assembly
graph. It supports both nucleotide and amino-acid pHMMs performing nt-to-aa
translation on-fly walking through frameshifts.</p>
<p><strong>PathRacer</strong> has two versions: main <code>pathracer</code> is for aliening both nucleotide and amino-acid pHMMs against assembly graphs and
<code>pathracer-seq-fs</code> is for aligning amino-acid pHMMs against separate sequences allowing indels in the nucleotide space.
<code>pathracer</code> is supposed to be used on complex metagenome assembly graphs for fragmented gens assembly and annotation.
<code>pathracer-fs-seq</code> is supposed to be used as a replacement of original <strong>HMMer</strong> for sequences with high indel rate, e.g., PacBio/ONT contigs.</p>
<p>Both tool use extended pHMM model allowing frame shifts:</p>
<p><img alt="Scheme of extended pHMM" src="pHMM_with_frameshifts.svg" /></p>
<p>but for <code>pathracer-seq-fs</code> this extension is crucial: for aligning amino-acid pHMMs without allowing indels in the nucleotide space
six frame translation + <code>hmmsearch</code> from <strong>HMMer</strong> package is more than enough.</p>
<h2 id="compilation">Compilation</h2>
<p>To compile PathRacer, run</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>./spades_compile -SPADES_ENABLE_PROJECTS=pathracer
</span></code></pre></div>
<p>After the compilation is complete, <code>pathracer</code> executable will be located in the <code>bin</code> folder.</p>
<h2 id="input">Input</h2>
<p>Currently, the tool supports only <em>de Bruijn</em> graphs in GFA format as produced by <strong>SPAdes</strong> or compatible assembler in this matter (e.g., <strong>MEGAHIT</strong>).
Contact us if you need some other format support. Input sequences are supposed to be in FASTA/FASTQ format.</p>
<p>Profile HMM should be in <strong>HMMer3</strong> format, but one can pass nucleotide or amino acid sequences as well.
These sequences will be converted to proxy pHMMs.
Aligning of these pHMMs would be equivalent to performing alignment using Levenshtein distance for each input sequence.</p>
<h2 id="pathracer-tool">pathracer tool</h2>
<h3 id="pathracer-command-line-options">pathracer command line options</h3>
<p>Required positional arguments:</p>
<ol>
<li>Query file (.hmm file or .fasta)</li>
<li>Assembly graph in GFA format</li>
</ol>
<p>Main options:</p>
<ul>
<li><code>--output</code>, <code>-o</code> DIR: output directory</li>
<li><code>--hmm</code> | <code>--nt</code> | <code>--aa</code>: perform match against pHMM(s) [default] | nucleotide sequences | amino acid sequences</li>
<li><code>--queries</code> Q1 [Q2 [...]]: queries names to lookup [default: all queries from input query file]</li>
<li><code>--edges</code> E1 [E2 [...]]: match around particular edges [default: all graph edges]</li>
<li><code>--global</code> | <code>--local</code>: perform HMM-global, graph-local (aka <em>glocal</em>, default) or HMM-local, graph-local HMM matching</li>
<li><code>--length</code>, <code>-l</code> L: minimal length of resultant matched sequence; if &le;1 then to be multiplied on aligned HMM length [default: 0.9]</li>
<li><code>--indel-rate</code>, <code>-r</code> RATE: expected rate of nucleotides indels in graph edges [default: 0]. Used for AA pHMM aligning with frameshifts</li>
<li><code>--top</code> N: extract up to <em>N</em> top scored paths [default: 10000]; only unique paths are reported and therefore</li>
<li><code>--rescore</code>: rescore resulting paths by <strong>HMMer</strong> and produce output tables in <strong>HMMer</strong> standard formats</li>
<li><code>--threads</code>, <code>-t</code> T: the total number of CPU threads to use [default: 16]</li>
<li><code>--parallel-components</code>: process connected components of neighborhood subgraph in parallel</li>
<li><code>--memory</code>, <code>-m</code> M: RAM limit in GB (<strong>PathRacer</strong> terminates if the limit is exceeded) [default: 100]</li>
<li><code>--annotate-graph</code>: emit paths in GFA graph</li>
</ul>
<p>Heuristics options:</p>
<ul>
<li><code>--max-size</code> MAX_SIZE: maximal component size to consider [default: INF]</li>
<li><code>--max-insertion-length</code>: maximal allowed number of successive I-emissions [default: 30]</li>
<li><code>--no-top-score-filter</code>: disable top score Event Graph vertices filter. Increases sensitivity of deep analysis (<code>--top</code> &gt; 50000)</li>
</ul>
<p>Debug output control:</p>
<ul>
<li><code>--debug</code>: enable extensive debug console output</li>
<li><code>--draw</code>: draw pictures around the interesting edges</li>
<li><code>--export-event-graph</code>: export Event Graph in binary format</li>
</ul>
<p><em>In addition:</em> Some other developer options that are not supposed to be tuned by the end-user. Could be removed in further releases.</p>
<h3 id="pathracer-output">pathracer output</h3>
<p>For each input pHMM (gene model) <code>pathracer</code> reports:</p>
<ul>
<li><strong>&lt;gene_name&gt;.seqs.fa</strong>: sequences correspondent to <em>N</em> best scored paths ordered by score along with their alignment in CIGAR format</li>
<li><strong>&lt;gene_name&gt;.nucls.fa</strong>: <em>(for amino acids pHHMs only)</em> the same sequences in nucleotides</li>
<li><strong>&lt;gene_name&gt;.edges.fa</strong>: unique graph edge paths sequences corresponding to best scored paths</li>
<li><strong>&lt;gene_name&gt;.{domtblout, pfamtblout, tblout}</strong>: <em>(optional)</em> edge paths realignment by <strong>HMMer</strong> in various default output formats</li>
<li><strong>event_graph_&lt;gene_name&gt;_component_&lt;component_id&gt;_size_&lt;component_size&gt;.bin</strong>: <em>(optional, debug output)</em> connected components of the event graph graph</li>
<li><strong>&lt;component_id&gt;.dot</strong>: <em>(optional, plot)</em> connected component of matched neighborhood subgraph</li>
<li><strong>&lt;component_id&gt;_&lt;path_index&gt;.dot</strong>: <em>(optional, plot)</em> neighborhood of the found path</li>
</ul>
<p>In addition:</p>
<ul>
<li><strong>all.edges.fa</strong>: unique edge paths for all pHMMs in one file</li>
<li><strong>pathracer.log</strong>: log file</li>
<li><strong>graph_with_hmm_paths.gfa</strong>: <em>(optional)</em> input graph with top scored paths added</li>
</ul>
<h2 id="pathracer-seq-fs-tool">pathracer-seq-fs tool</h2>
<h3 id="pathracer-seq-fs-command-line-options">pathracer-seq-fs command line options</h3>
<p>Required positional arguments:</p>
<ol>
<li>Query .hmm file (.fasta is not supported yet)</li>
<li>Sequence file (.fasta or .fastq)</li>
</ol>
<p>Main options:</p>
<ul>
<li><code>--output</code>, <code>--queries</code>, <code>--global | --local</code>, <code>--top</code>, <code>--threads</code>, <code>--memory</code>: the same as in main <code>pathracer</code></li>
<li><code>--sequences</code> S1 [S2 [...]]: sequence IDs to process [default: all input sequences]</li>
<li><code>--indel-rate</code>, <code>-r</code> RATE: expected rate of nucleotides indels in graph edges [default: 0.05]. Used for AA pHMM aligning with frameshifts</li>
<li><code>--max-fs</code> N: maximal allowed number of frameshifts in a reported sequence [default: 10]</li>
<li><code>--cutoff</code> CUTOFF: bitscore cutoff for reported match; if &lt;= 1 then to be multiplied on GA HMM cutoff [default: 0.7]",</li>
</ul>
<p>Heuristics options:
<em>The same as in main <code>pathracer</code></em></p>
<h3 id="pathracer-seq-fs-output">pathracer-seq-fs output</h3>
<p>For each input pHMM (gene model): <strong>&lt;gene_name&gt;.seqs.fa</strong> and <strong>&lt;gene_name&gt;.nucls.fa</strong>, the same as in main <code>pathracer</code></p>
<h2 id="output-files-format">Output files format</h2>
<p><strong>&lt;gene_name&gt;.seqs.fa</strong> and <strong>&lt;gene_name&gt;.nucls.fa</strong> files contain metainformation in FASTA headers.
For main <code>pathracer</code> the header format is:</p>
<blockquote>
<p><code>
<span>&gt;</span>Score=<em>PathRacer score</em>|Edges=<em>edges path</em>|Position=<em>starting position on the first edge</em>|Alignment=<em>CIGAR alignment</em>
</code></p>
</blockquote>
<p>E.g.:</p>
<blockquote>
<p><code>
<span>&gt;</span>Score=366.239|Edges=255162_24353'|Position=9210|Alignment=186M2D186M
</code></p>
</blockquote>
<p>Prime (') after an edge ID means reverse complement</p>
<p>For <code>pathracer-seq-fs</code> the header format is:</p>
<blockquote>
<p><code>
<span>&gt;</span>Score=<em>PathRacer score</em>|Bitscore=<em>HMMer bitscore for the whole sequence without incomplete codons</em>|PartialBitscore=<em>Maximal HMMer bitscore for fragment between frameshifts</em>|Seq=<em>Sequence ID</em>|Position=<em>Starting position in the sequence</em>|Frameshifts=<em>#Frameshifts</em>|Alignment=<em>CIGAR alignment</em>
</code></p>
</blockquote>
<p>E.g.</p>
<blockquote>
<p><code>
<span>&gt;</span>Score=342.689|Bitscore=539.274|PartialBitscore=238.41|Seq=RB12-N|Position=2935|Alignment=55M1G1M1D20M1D14M1I3M2D11M1P1M1D64M1D62M1D1M1G23M1D30M<br />
MSLYRRLVLLSCLSWPLAGFSATALTNLVAEPFAKLEQDFGGSIGVYAMDTGSGA=CSYR<br />
AEERFPLCSSFKGFLAAVLARSQQGRLAGHTHPLRQNALVPWSPIS-KYLTTGMTVAELS<br />
AAAVQYSDNAAANLLLKELGGPAGLTAFMRSIGDTTFRLDRWELELNSAIRAMRAIPHRR<br />
ARDGKLTKLTLGSALAAPQRQQFVDWLKGNTTGNHRIRAAVPADWAVGDKTGTCG=YGTA<br />
NDYAVVWPTGRAPIVLAVYRAPNKDDKHSEAVIAAAARLALEDWASTAV
</code></p>
</blockquote>
<p>For alignment with frameshifts the extemded CIGAR/FASTA is used:
P/"-" &mdash; one nucleotide insertion, G/"=" &mdash; two nucleotides insertion</p>
<h2 id="examples">Examples</h2>
<p>One can download example datasets from <a href="http://cab.spbu.ru/software/pathracer/">http://cab.spbu.ru/software/pathracer/</a></p>
<ul>
<li><strong>urban_strain.gfa</strong>: strain assembly graph of Singapore clinical isolation ward wastewater metagenome (SRA accession SRR5997548, dataset H1)</li>
<li><strong>bla_all.hmm</strong>: 159 beta-lactamase family pHMMs from <strong>AMRFinder</strong> database <a href="https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/">https://www.ncbi.nlm.nih.gov/pathogens/antimicrobial-resistance/AMRFinder/</a></li>
<li><strong>ecoli_mc.gfa</strong>: assembly graph of <em>E. coli str. K12 substr. MG1655</em> multicell isolate (SRA accession ERA000206)</li>
<li><strong>bac.hmm</strong>: <em>16S</em>, <em>23S</em>, and <em>5S</em> ribosomal subunit pHMMs</li>
<li><strong>synth_strain_gbuilder.gfa</strong>: strain assembly graph of SYNTH mock metagenome dataset (SRA accession SRR606249)</li>
<li><strong>synth16S_new.fa</strong>: <em>16S</em> rRNA sequences of SYNTH organisms obtained from <a href="https://www.arb-silva.de/">SILVA</a> database</li>
<li><strong>plasmids-ONT.fa</strong>: 7 ONT-assembled AMR-related plasmids from <a href="https://www.ncbi.nlm.nih.gov/pubmed/29325009">Li et al, 2018</a></li>
</ul>
<p>Lookup for beta-lactamase genes (amino acid pHMMs) in Singapore wastewater<br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>pathracer bla_all.hmm urban_strain.gfa --output pathracer_urban_strain_bla_all
</span></code></pre></div></p>
<p>Lookup for beta-lactamase genes (amino acid pHMMs) in AMR ONT plasmids (many indels!) <br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>pathracer-fs-seq bla_all.hmm plasmids-ONT.fa --output pathracer_plasmids_ont_bla_all
</span></code></pre></div></p>
<p>Lookup for <em>16S</em>/<em>5S</em>/<em>23S</em> (nucleotide HMMs) in <em>E.coli</em> multicell assembly<br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>pathracer bac.hmm ecoli_mc.gfa --output pathracer_ecoli_mc_bac
</span></code></pre></div></p>
<p>Lookup for known <em>16S</em> sequences in <em>E.coli</em> multicell assembly<br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>pathracer synth16S_new.fa ecoli_mc.gfa --nt --output pathracer_ecoli_mc_16S_seqs
</span></code></pre></div></p>
<p>Lookup for known <em>16S</em> sequences in SYNTH mock metagenome assembly<br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>pathracer synth16S_new.fa synth_strain_gbuilder.gfa --nt --output pathracer_synth_strain_gbuider_16S_seqs
</span></code></pre></div></p>
<p>Let us extract <strong>all</strong> <em>16S</em> sequences from SYNTH mock metagenome assembly.
For this we increase <code>--top</code> and disable Event Graph vertices filter (<code>--no-top-score-filter</code>)
Deep analysis of extremely complicated dataset also require stack and memory limits tuning<br />
<div class="language-text highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>ulimit -s unlimited &amp;&amp;  
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>export OMP_STACKSIZE=1G  
</span><span id="__span-6-3"><a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a>pathracer bac.hmm synth_strain_gbuilder.gfa --queries 16S_rRNA -m 250 --top 1000000 --output pathracer_synth_strain_gbuilder_16s --no-top-score-filter
</span></code></pre></div></p>
<h2 id="references">References</h2>
<p>If you are using <strong>PathRacer</strong> in your research, please cite:</p>
<p><a href="https://link.springer.com/chapter/10.1007/978-3-030-18174-1_6">Shlemov and Korobeynikov, 2019</a></p>
<p>In case of any problems running <strong>PathRacer</strong> please contact <a href="https://github.com/ablab/spades/issues">SPAdes support</a> attaching the log file.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": ".", "features": ["content.code.copy"], "search": "assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>