
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="pathracer.html">
      
      
        <link rel="next" href="standalone.html">
      
      
      <link rel="icon" href="spades.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.12">
    
    
      
        <title>Sequence to graph alignment - SPAdes Assembly Toolkit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.2afb09e1.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#spaligner-long-read-to-graph-aligner" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="SPAdes Assembly Toolkit" class="md-header__button md-logo" aria-label="SPAdes Assembly Toolkit" data-md-component="logo">
      
  <img src="spades.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            SPAdes Assembly Toolkit
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Sequence to graph alignment
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="SPAdes Assembly Toolkit" class="md-nav__button md-logo" aria-label="SPAdes Assembly Toolkit" data-md-component="logo">
      
  <img src="spades.png" alt="logo">

    </a>
    SPAdes Assembly Toolkit
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="index.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="getting-started.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quick start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="installation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    SPAdes input data
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            SPAdes input data
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="input.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Basic input
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="hybrid.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Hybrid assembly
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="datatypes.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tips on the parameters
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="running.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command line options
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="output.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SPAdes output
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="hmm.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    HMM-guided mode
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="rna.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Transcriptome assembly
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="hpc.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Running SPAdes across multiple nodes of a cluster
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="binspreader.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Binning refining
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="pathracer.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    HMM mapping on assembly graph
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Sequence to graph alignment
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="spaligner.html" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Sequence to graph alignment
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#compilation" class="md-nav__link">
    <span class="md-ellipsis">
      Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-spaligner" class="md-nav__link">
    <span class="md-ellipsis">
      Running SPAligner
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Running SPAligner">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      Command line options
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#results-interpretation" class="md-nav__link">
    <span class="md-ellipsis">
      Results interpretation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#algorithm-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters-tuning" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters tuning
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Parameters tuning">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#general-alignment-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      General alignment parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#internal-run-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Internal run parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="standalone.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    SPAdes tools
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="citation.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Citation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="feedback.html" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Feedback
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#compilation" class="md-nav__link">
    <span class="md-ellipsis">
      Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-spaligner" class="md-nav__link">
    <span class="md-ellipsis">
      Running SPAligner
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Running SPAligner">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      Command line options
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#results-interpretation" class="md-nav__link">
    <span class="md-ellipsis">
      Results interpretation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#algorithm-overview" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters-tuning" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters tuning
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Parameters tuning">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#general-alignment-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      General alignment parameters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#internal-run-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Internal run parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      References
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="spaligner-long-read-to-graph-aligner">SPAligner: long read to graph aligner</h1>
<p>SPAligner is a tool for fast and accurate alignment of nucleotide sequences to assembly graphs.
It takes file with sequences (in fasta/fastq format) and assembly in GFA format and outputs long read
to graph alignment in various formats (such as tsv, fasta and <a href="https://github.com/ocxtal/gpa" title="GPA-format spec">GPA</a>).</p>
<h2 id="compilation">Compilation</h2>
<p>To compile SPAligner, run</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>./spades_compile -SPADES_ENABLE_PROJECTS=spaligner
</span></code></pre></div>
<p>After the compilation is complete, <code>spaligner</code> executable will be located in the <code>bin</code> folder.</p>
<h2 id="running-spaligner">Running SPAligner</h2>
<p>Synopsis: </p>
<div class="language-text highlight"><pre><span></span><code>spaligner spaligner_config.yaml \    # config file
          -d pacbio \                # data type: pacbio, nanopore
          -g assembly_graph.gfa \    # assembly graph 
          -k 77 \                    # graph k-mer size
          -s pacbio_reads.fastq.gz \ # input sequences / reads
          -t 8                       # number of threads
</code></pre></div>
<p>By default, <code>spaligner_config.yaml</code> can be found in <code>src/projects/spaligner/</code>.</p>
<p>Alignments will be saved to <code>spaligner_result/alignment.tsv</code> by default.</p>
<h3 id="command-line-options">Command line options</h3>
<p><code>-d &lt;type&gt;</code>
    long reads type: <code>nanopore</code> or <code>pacbio</code></p>
<p><code>-s &lt;filename&gt;</code>
    file with sequences in FASTA or FASTQ formats (can be gzipped)</p>
<p><code>-g &lt;filename&gt;</code>
    file with an assembly graph in GFA format</p>
<p><code>-k &lt;int&gt;</code>
    k-mer length that was used for graph construction</p>
<p><code>-t &lt;int&gt;</code>
    number of threads (default: 8)</p>
<p><code>-o, --outdir &lt;dir&gt;</code>
    output directory to use (default: <code>spaligner_result/</code>)</p>
<h2 id="output">Output</h2>
<p>SPAligner can represent the results in three formats: <em>.tsv (default), </em>.fasta and <a href="https://github.com/ocxtal/gpa" title="GPA-format spec">*.gpa</a>.</p>
<div class="language-text highlight"><pre><span></span><code>spaligner_result/alignment.tsv              tab-separated file with alignments information, each line represents an alignment of a single sequence
spaligner_result/alignment.fasta            each record represents alignment of a sequence onto assembly graph
spaligner_result/alignment.gpa              alignment stored in gpa-format
</code></pre></div>
<h2 id="results-interpretation">Results interpretation</h2>
<p>SPAligner can represent the results in three formats: <em>.tsv (default), </em>.fasta and <a href="https://github.com/ocxtal/gpa" title="GPA-format spec">*.gpa</a>.
Name of each record in fasta files shows information about alignment position on graph and on sequence. </p>
<p><strong>Example 1</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>&gt;name|Edges=1-,2+,5+|start_g=3283|end_g=35|start_s=0|end_s=291
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>ATGAAAATCACTCCTGAACAGGCTCGTGAGGCTCTGGATGCCTGGATATGTCGACCAGGAATGACACAGGAGCAGGCGACGATATTAATCACTGAAGCATTCTGGGCTTTGAAAGAGCGCCCGAACATCGATGTTCAGCGTGTCACATATGAAGGTGGCGCGATTGATCAGCGAGCGCTTGGCGTTAATCGAGTGAAGATATTTGAACGCTGGAAGGCTATCGACACCAGGGATAAGC
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>GTGAAAAGTTCACGGCGCTAGTGCCTGCAATTATGGAGGCTACCACTGGATGA
</span></code></pre></div>
<p>name — sequence name<br/>
1-,2+,5+ — alignment path<br/>
3283 — start position of alignment on the first edge of the path (here on conjugate edge to edge with id=1)<br/>
35 —  end position of alignment on the last edge of the path (here on edge with id=5)<br/>
0 — start position of alignment on sequence<br/>
291 — end position of  alignment on sequence<br/></p>
<p>Each line in tsv-file represents alignments of a single read.</p>
<p><strong>Example 2</strong></p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>name    0      2491  536  1142  2491    44+,24+,22+,1+,38-  909,4,115,1,1142    AAACTTTTATTGTGCATACGGCGATTAAGACGGGAAAAGTCGGTGAT...
</span></code></pre></div>
<p>name — sequence name<br/>
0 — start position of alignment on sequence<br/>
2491 — end position of  alignment on sequence<br/>
536 — start position of alignment on the first edge of the Path (here on edge with id=44)<br/>
1142 —  end position of alignment on the last edge of the Path (here on conjugate edge to edge with id=38)<br/>
2491 — sequence length<br/>
44+,24+,22+,1+,38- — Path of the alignment <br/>
909,4,115,1,1142 — lengths of the alignment on each edge of the Path respectively (44+,24+,22+,1+,38-) <br/>
AGGTTGTTTTTTGTTTCTTCCGC... — sequence of alignment Path <br/></p>
<p><strong>Example 3</strong><br/></p>
<p>Sometimes sequence alignment on the graph can be represented as several non-overlapping subpaths (if there is no alignment with appropriate score between two consecutive bwa anchors). <br/>
So, there can be several unconnected alignments of sequence onto assembly graph and several start positions, end positions, paths etc.:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>name     4,10  7,19       2,7   5,6       19    123+;288-,128+       3;3,6 GAT;TTATCCGGG
</span></code></pre></div>
<p>The sequence <em>name</em> has two alignments on the graph:</p>
<ol>
<li>The first alignment starts on sequence on position 4 and ends on position 7.
Corresponding path consists of a single edge 123+ (i.e. 123+) with start on position 2 and end on position 5.
Path sequence: GAT.</li>
<li>While the second alignment covers the end of the sequence, starting on sequence on position 10 and ending on position 19. 
Corresponding path consists of two edges 288- and 128+ (i.e. 288-,128+) with start on position 7 (on 288-) and end on position 6 (on 128+).
Path sequence: TTATCCGGG.</li>
</ol>
<p>If a sequence was not fully aligned, SPAligner tries to prolong the longest alignment subpath in order to reconstruct a full alignment path. In <strong>Example 2</strong> SPAligner was not able to prolong any of two given alignments.</p>
<h2 id="algorithm-overview">Algorithm overview</h2>
<p>Overview of the alignment of the nucleotide query sequence <em>S</em> (orange bar) to assembly graph <em>G</em>. Assembly graph edges are considered directed left-to-right (explicit edge orientation was omitted to improve the clarity).</p>
<p><img alt="pipeline" src="spaligner.jpg" /></p>
<ol>
<li><strong>Anchor search.</strong> Anchors (regions of high similarity) between the query and the edge labels are identified with <a href="http://bio-bwa.sourceforge.net/">BWA-MEM</a>. </li>
<li><strong>Anchor filtering.</strong> Anchors shorter than <em>K</em>, assembly graph <em>K</em>-mer size,(anchors 2, 6, 11), anchors “in the middle” of long edge (anchor 7) or ambiguous anchors (anchor 10 mostly covered by anchor 9, both anchors 4 and 5) are discarded.</li>
<li><strong>Anchor chaining.</strong> Heaviest chain of compatible anchors (chain 3-&gt;8-&gt;9) is determined.</li>
<li><strong>Reconstruction of filling paths.</strong> Paths for fragments of the query between the consecutive chain anchors (as well as left- and right-most fragments) are reconstructed. The procedure is performed using fast library for sequence alignment <a href="https://github.com/Martinsos/edlib">Edlib</a>.</li>
</ol>
<h2 id="parameters-tuning">Parameters tuning</h2>
<p>Full list of parameters can be found in spaligner_config.yaml. </p>
<h3 id="general-alignment-parameters">General alignment parameters</h3>
<ul>
<li><code>run_dijkstra: true</code> Run Dijkstra algorithm to find alignment between anchors, if <code>run_dijkstra: false</code>, SPAligner will check limited number of paths and return the best one.</li>
<li>
<p><code>restore_ends: true</code> Restore alignment path before leftmost anchor and after rightmost anchor.</p>
</li>
<li>
<p><code>internal_length_cutoff: 200</code> Anchors with length &lt; <code>internal_length_cutoff</code> will be filtered out.</p>
</li>
<li><code>path_limit_stretching: 1.3</code> Pair of anchors is considered to be compatible if (the minimal distance between them in graph) 
                                &lt; <code>path_limit_stretching</code> * (the distance between their positions on sequence).</li>
<li><code>path_limit_pressing: 0.7</code> Pair of anchors is considered to be compatible if (the minimal distance between them in graph) 
                                &gt; <code>path_limit_pressing</code> * (the distance between their positions on sequence).</li>
<li><code>max_path_in_chaining: 15000</code> Limit on number of paths to consider between two anchors on anchors chaining step.</li>
<li><code>max_vertex_in_chaining: 5000</code> Limit on number of vertices to consider between two anchors on anchors chaining step.</li>
</ul>
<h3 id="internal-run-parameters">Internal run parameters</h3>
<ul>
<li><code>queue_limit: 1000000</code> Limit on queue length. </li>
<li><code>iteration_limit: 1000000</code> Limit on total number of queue extraction. </li>
<li><code>updates_limit: 1000000</code> Limit on number of updates of shortest distance for all states.</li>
<li><code>find_shortest_path: true</code> If <code>find_shortest_path: false</code> Dijkstra algorithm will stop when it reaches finish state without searching for shortest path.</li>
<li><code>restore_mapping: false</code> If <code>restore_mapping: true</code> Dijkstra algorithm will return full alignment information (used by for developers).</li>
<li><code>penalty_ratio: 0.1</code> Algorithm never considers states representing an alignment of a prefix S[0:i] with score more than min_score(i) + i*penalty_ratio (for nucleotide sequence alignment only).</li>
<li><code>max_ed_proportion: 3</code> Maximal edit distance is bounded by a fraction of the query sequence length |S|/max_ed_proportion. Increase of max_ed_proportion leads to shorter alignments but with higer identity.</li>
<li><code>ed_lower_bound: 500</code> Minimal penalty score of alignment.</li>
<li><code>ed_upper_bound: 2000</code> Maximal penalty score of alignment.</li>
<li><code>max_gs_states: 120000000</code> If number of queue states exceeds <code>max_gs_limit</code> then shortest path search is not performed (for nucleotide sequence alignment only).</li>
<li><code>max_restorable_length: 5000</code> If distance between two anchors or between leftmost/rightmost anchor and start/end exceeds <code>max_restorable_length</code> then shortest path search is not performed.</li>
</ul>
<p>Increase of <code>max_gs_states</code>, <code>max_restorable_length</code>, <code>queue_limit</code>, <code>iteration_limit</code> or <code>updates_limit</code> may lead to longer alignments with the same identity level, but slows down the process and can use much more memory. Please change them if you 100% confident in what you are doing.</p>
<p>Turning off restore_ends or run_dijkstra in nucleotide sequence alignment mode leads to shorter alignments, but considerable speed-up.</p>
<h2 id="references">References</h2>
<p>If you are using <strong>SPAligner</strong> in your research, please cite:</p>
<p><a href="https://link.springer.com/article/10.1186/s12859-020-03590-7">Dvorkina et al., 2020</a></p>
<p>For any questions or suggestions please do not hesitate to contact Tatiana Dvorkina <a href="&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#116;&#101;&#100;&#118;&#111;&#114;&#107;&#105;&#110;&#97;&#64;&#103;&#109;&#97;&#105;&#108;&#46;&#99;&#111;&#109;">&#116;&#101;&#100;&#118;&#111;&#114;&#107;&#105;&#110;&#97;&#64;&#103;&#109;&#97;&#105;&#108;&#46;&#99;&#111;&#109;</a>.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": ".", "features": ["content.code.copy"], "search": "assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>