<title>SPAdes Genome Assembler changelog</title>
<body>

<h2>SPAdes Genome Assembler changelog</h2>


<h3>SPAdes 3.15.5, 14 July 2022</h3>

<p>NEW: Support DNA HMMs. <p>

<p>FIX: use-after-free in PB aligner. <p>

<p>FIX: duplicated sequences in metaplasmid / metaviral mode. <p>

<h3>SPAdes 3.15.4, 1 February 2022</h3>

<p>FIX: MacOS Monterey memory limit failure. <p>

<p>FIX: upgrade pyyaml to run correctly with Python 3.10. <p>

<p>FIX: WSL for py2 check. <p>

<p>FIX: A few stability fixes.<p>

<h3>SPAdes 3.15.3, 22 July 2021</h3>

<p>FIX: trusted contigs failure.</p>

<p>FIX: clarification & refining the output of bgcSPAdes and coronaSPAdes.</p>

<p>FIX: usage of >9 libraries in a single SPAdes run.</p>

<p>FIX: improvements in <code>spades-read-filter</code> tool.</p>

<h3>SPAdes 3.15.2, 8 March 2021</h3>

<p>FIX: meta-viral pipeline bugs.</p>

<p>FIX: coronaspades.py wrapper, copy proper files to the output folder.</p>

<p>FIX: coronaSPAdes instability</p>

<h3>SPAdes 3.15.1, 18 February 2021</h3>

<p>FIX: Gap closer failure when using multiple libraries.</p>

<p>FIX: Gap closer excessive memory consumption.</p>

<p>IMPROVE: coronaSPAdes output.</p>

<h3>SPAdes 3.15.0, 11 January 2021</h3>

<p>NEW: CoronaSPAdes pipeline for assembly of full-length coronaviridae genomes from the transcriptomic and metatranscriptomic data.</p>

<p>NEW: Meta-Viral and RNA-Viral pipelines for identifying viral genomes for metagenomic and metatranscriptomic data.</p>

<p>NEW: Novel algorithm for trusted contig usage.</p>

<p>NEW: Switched to <a href="https://github.com/microsoft/mimalloc">mimalloc</a> memory allocator.</p>

<p>NEW: PlasmidSPAdes and bgcSPAdes now support assembly graph as an input.</p>

<p>CHANGE: Significant improvements and fixes for metaplasmid pipeline.</p>

<p>CHANGE: Multiple performance improvements in simplification and repeat resolving procedures.</p>

<p>DEPRECATED: Support for Lucigen NxSeqÂ® Long Mate Pair reads.</p>

<p>DEPRECATED: truSPAdes pipeline for TruSeq barcode assembly (still present in this release but no longer supported).</p>

<h3>SPAdes 3.14.1, 27 April 2020</h3>

<p>FIX: metaplasmidSPAdes contig output.</p>

<p>FIX: read filtering binary.</p>

<p>FIX: biosyntheticSPAdes pipeline fixed.</p>

<p>FIX: fixed truSPAdes for Python 3.6+.</p>

<p>FIX: bug in the internal mismatch correction procedure.</p>

<p>FIX: Soft and hard-filtered transcripts are now copied to the output folder in rnaSPAdes.</p>

<p>FIX: Several usability fixes in `spades.py'.</p>

<p>FIX: meta-plasmid options added to the manual.</p>

<p>FIX: several minor fixes in the user manual.</p>

<h3>SPAdes 3.14.0, 27 December 2019</h3>

<p>NEW: BiosyntheticSPAdes pipeline for predicting Biosynthetic Gene Clusters.</p>

<p>NEW: Hybrid transcriptome assembly with rnaSPAdes.</p>

<p>NEW: Plasmid detection from metagenomic samples.</p>

<p>NEW: Special <code>--isolate</code> option for assembly of standard datasets with good coverage (>100x).</p>

<p>NEW: Standalone tool for reads filtration based on k-mer coverage.</p>

<p>NEW: Standalone tool for estimating approximate number of unique k-mers in reads.</p>

<p>CHANGE: Improved SPAligner tool.</p>

<p>CHANGE: Reworked python code, faster sequence transfer between different k-mer stages.</p>

<p>CHANGE: Multiple performance improvements in graph construction and simplification procedures.</p>

<p>FIX: BWA aligner failure for large graphs.</p>

<p>FIX: Failure when additional paired-end libraries with reads shorter than final k-mer length are provided.</p>

<h3>SPAdes 3.13.2, 31 October 2019</h3>

<p>FIX: Incorrect k-mer size estimation in rnaSPAdes.</p>

<h3>SPAdes 3.13.1, 11 April 2019</h3>

<p>CHANGE: Removed BayesHammer from rnaSPAdes pipeline.</p>

<p>CHANGE: Improved rnaSPAdes performance on large datasets.</p>

<p>FIX: Failure in contig output in rnaSPAdes.</p>

<h3>SPAdes 3.13.0, 11 October 2018</h3>

<p>CHANGE: Switched to multi-k-mer mode in rnaSPAdes, k-mer values are detected automatically based on read length.</p>

<p>CHANGE: Added manual as README.md in markdown format for github.</p>

<p>FIX: Updated BWA and switched to RopeBWT, which allows to handle large graphs.</p>

<p>FIX: Assert for path.length() > 0 in rnaSPAdes. </p>

<p>FIX: CQF and k-mer counting. </p>

<h3>SPAdes 3.12.0, 14 May 2018</h3>

<p>NEW: Support for merged paired-end reads.</p>

<p>NEW: Experimental pipeline for metagenome hybrid assemblies.</p>

<p>NEW: Standalone graph builder application.</p>

<p>NEW: Standalone k-mer counting application.</p>

<p>NEW: Standalone long read to graph aligner.</p>

<p>CHANGE: Significant improvements in hybrid assembly pipeline.</p>

<p>CHANGE: Faster read alignment using BWA.</p>

<p>CHANGE: Improvements in metaSPAdes results.</p>

<p>CHANGE: More sensitive results for rnaSPAdes.</p>

<p>CHANGE: All binaries for SPAdes pipeline steps now have <code>spades-</code> prefix in its name.</p>

<p>CHANGE: Better running time and RAM consumption for graph construction stage.</p>

<p>CHANGE: Overall performance improvements.</p>

<p>FIX: K value estimation for rnaSPAdes.</p>

<p>DEPRECATED: dipSPAdes pipeline for highly polymorphic diploid genomes (still present in the release but no longer supported).</p>

<h3>SPAdes 3.11.1, 1 October 2017</h3>

<p>FIX: Handling spaces in path during mismatch correction.</p>

<p>FIX: Python3 support in rnaSPAdes.</p>

<p>FIX: K value estimation for long reads.</p>

<p>FIX: Processing long reads alignments.</p>

<h3>SPAdes 3.11.0, 1 September 2017</h3>

<p>NEW: Support for strand-specific RNA-Seq data in rnaSPAdes.</p>

<p>NEW: Coverage based isoform detection in rnaSPAdes.</p>

<p>NEW: Reworked IonHammer read error correction module.</p>

<p>CHANGE: Improved tandem repeat resolution accuracy.</p>

<p>CHANGE: Better performance of exSPAnder module.</p>

<p>CHANGE: metaSPAdes pipeline improvements.</p>

<p>CHANGE: Better running time and RAM consumption for the entire pipeline.</p>

<p>FIX: Incomplete paths in GFA output.</p>

<p>FIX: Mismatch and indel rate in careful mode for isolate datasets (esp. low covered ones).</p>

<p>FIX: Occasional hanging of edge disconnection procedure in metaSPAdes.</p>

<h3>SPAdes 3.10.1, 1 March 2017</h3>

<p>FIX: Build for MacOS.</p>

<p>FIX: Minor bugs in hybridSPAdes pipeline.</p>

<p>FIX: <code>--continue</code> option for metaSPAdes.</p>

<p>FIX: <code>--tmp-dir</code> is now works correctly for MismatchCorrector.</p>

<p>FIX: <code>Assertion `overlap <= k_' failed</code> in rnaSPAdes and metaSPAdes.</p>

<p>FIX: <code>Assertion `path.Length() > 0' failed</code> in metaSPAdes.</p>

<h3>SPAdes 3.10.0, 27 January 2017</h3>

<p>NEW: Scaffolding algorithm for mate-pairs and long reads.</p>

<p>NEW: Contigs and graph output in GFA format.

<p>CHANGE: Better running time and RAM consumption for all pipelines.</p>

<p>CHANGE: Improvements in metagenomic pipeline.</p>

<p>CHANGE: Improved isoform detection algorithm in rnaSPAdes.</p>

<h3>SPAdes 3.9.1, 4 December 2016</h3>

<p>FIX: macOS Sierra crash.</p>

<h3>SPAdes 3.9.0, 23 July 2016</h3>

<p>NEW: rnaSPAdes pipeline for de novo transcriptome assembly from RNA-Seq data.</p>

<p>CHANGE: Improved memory consumption in metagenomic pipeline.</p>

<p>FIX: Several minor bugs.</p>

<h3>SPAdes 3.8.2, 10 July 2016</h3>

<p> FIX: Several minor bug-fixes for metaSPAdes and SPAdes pipelines.</p>

<h3>SPAdes 3.8.1, 8 June 2016</h3>

<p>FIX: plasmidSPAdes now works with PacBio/Nanopore reads.</p>

<h3>SPAdes 3.8.0, 1 June 2016</h3>

<p>NEW: Added plasmidSPAdes &ndash;  a pipeline designed for extracting and assembling plasmids from WGS data sets.</p>

<p>CHANGE: Significant improvements in metaSPAdes performance.</p>

<p>CHANGE: Improved running time and RAM consumption.</p>

<h3>SPAdes 3.7.1, 8 March 2016</h3>

<p>FIX: MismatchCorrector fixed for MaxOS.</p>

<h3>SPAdes 3.7.0, 24 February 2016</h3>

<p>NEW: metaSPAdes metagenomic pipeline.</p>

<p>CHANGE: improved performance for both error correction and assembly stages.</p>

<p>FIX: Multiple bug fixes.</p>

<h3>SPAdes 3.6.2, 20 November 2015</h3>

<p>NEW: Contigs/scaffolds paths for assembly_graph.fastg in Bandage-supported format.</p>

<p>FIX: Multithreaded MismatchCorrector.</p>

<p>FIX: BayesHammer bug fixes.</p>

<p>FIX: Python 3.5 support; python 3 support for truSPAdes.</p>

<h3>SPAdes 3.6.1, 4 October 2015</h3>

<p>CHANGE: No misleading FASTG files, only assembly graph is saved in FASTG format.

<p>FIX: Multiple bugfixes.</p>

<h3>SPAdes 3.6.0, 17 August 2015</h3>

<p>NEW: Added truSPAdes &ndash; an assembler for short reads produced by Illumina TruSeq Long Read technology.</p>

<p>CHANGE: Better running time, less RAM consumption and improved results for BayesHammer error correction module.</p>

<p>CHANGE: Improvements and bugfixes in repeat resolution and scaffolding modules.</p>

<p>CHANGE: Improvements and bugfixes in dipSPAdes.</p>

<p>CHANGE: MismatchCorrector now uses bwa-mem.</p>

<p>FIX: Bugfixes in MismatchCorrector.</p>

<h3>SPAdes 3.5.0, 7 December 2014</h3>

<p>NEW: New MismatchCorrector module.</p>

<p>NEW: Support for Oxford Nanopore long reads.</p>

<p>NEW: Support for Lucigen NxMate mate-pair libraries.</p>

<p>NEW: Possibility to specify coverage cutoff: automatic and manual.</p>

<p>CHANGE: Better running time.</p>

<p>CHANGE: Improved RAM consumption.</p>

<p>CHANGE: High-quality mate-pairs are now assumed to have forward-revers orientation (same as paired-end).</p>

<p>FIX: Fixed FASTG format.</p>

<h3>SPAdes 3.1.1, 27 May 2014</h3>

<p>FIX: Several improvements in IonHammer.</p>

<p>FIX: Fixed a few minor bugs in repeat resolution and scaffolding.</p>

<h3>SPAdes 3.1.0, 27 May 2014</h3>

<p>NEW: Mate-pair only assembly with high-quality libraties.</p>

<p>NEW: Support for BAM files.</p>

<p>CHANGE: Improved IonTorrent pipeline.</p>

<p>CHANGE: Better quality and higher performance when using mate-pairs.</p>

<p>FIX: Fixed dipSPAdes bugs and user interface.</p>

<h3>SPAdes 3.0.0, 29 December 2013</h3>

<p>NEW: Module for assemblying diplod highly polymorphic genomes.</p>

<p>NEW: Support for PacBio reads.</p>

<p>NEW: Support for IonTorrent reads.</p>

<p>NEW: Support for Sanger reads and additional contigs.</p>

<p>NEW: Possibility to restart SPAdes starting from the specified check-point with the <code>--restart-from</code> option.</p>

<p>NEW: Output contigs/scaffolds in FASTA and FASTG.</p>

<p>CHANGE: Improved algorithm for mate-pair repeat resolution and scaffolding.</p>

<p>CHANGE: Improved N50 and misassembly rate for single-cell data sets with low genome fraction.</p>

<p>FIX: User-friendly handling for errors in mismatch corrector.</p>

<p>REMOVE: Rectangle graph repeat resolution module.</p>

<h3>SPAdes 2.5.1, 10 September 2013</h3>

<p>NEW: Python 3.2 and 3.3 compatibility.</p>

<p>NEW: Possibility to continue SPAdes run starting from the last check-point with the <code>--continue</code> option.</p>

<p>CHANGE: Decreased memory consumption for error correction module.</p>

<p>CHANGE: Improved misassembly rate for single-cell data sets with low genome fraction.</p>

<p>FIX: User-friendly handling for the case when paired reads do not align to the assembly graph.</p>

<h3>SPAdes 2.5.0, 2 July 2013</h3>

<p>NEW: Multiple paired-end and mate-pair libraries.</p>

<p>NEW: Recipe for assembling Illumina 2x150bp and 2x250bp reads.</p>

<p>CHANGE: Improved mismatch and indel rate.</p>

<h3>SPAdes 2.4.0, 26 February 2013</h3>

<p>NEW: Mismatch correction post-processing module.</p>

<p>NEW: Rectangle graph repeat resolution module as an option.</p>

<p>NEW: Build for Mac OS.</p>

<p>CHANGE: Improved assembly quality of standard (isolate) data sets.</p>

<p>CHANGE: Decreased memory consumption in error correction module (14 Gb instead of 24 Gb on E.coli test dataset).</p>

<p>REMOVE: SAM-file generation.</p>

<h3>SPAdes 2.3.0, 30 October 2012</h3>

<p>NEW: Generate scaffolds alongside with contigs.</p>

<p>CHANGE: Use N instead of A, C, G, T for the variations in repeats.</p>

<p>CHANGE: Memory requirements for E.coli test dataset decreased from 35 Gb to 24 Gb of RAM.</p>

<p>CHANGE: output_dir is a required command line parameter instead of project_name.</p>

<p>CHANGE: Simplified output directory structure.</p>

<p>CHANGE: CMake 2.8 is required instead of 2.6.</p>

<p>REMOVE: No dependency from boost library.</p>

<h3>SPAdes 2.2.1, 20 August 2012</h3>

<p>FIX: Avoid <code>Verification of expression 'v1 == conjugate(v2)' failed</code> error.</p>


<h3>SPAdes 2.2.0, 2 August 2012</h3>

<p>NEW: No special binaries for different K values.</p>

<p>NEW: Great improvements in error correction tool BayesHammer.</p>

<p>CHANGE: Memory requirements for E.coli test dataset decreased from 85 Gb to 35 Gb of RAM.</p>

<p>CHANGE: Only 1 iteration of BayesHammer by default.</p>

<p>NEW: Improved assembly quality.</p>

<h3>SPAdes 2.1.0, 28 May 2012</h3>

<p>NEW: Support multi-threading.</p>

<p>NEW: Improved algorithms.</p>

<p>NEW: Command-line interface.

<p>CHANGE: Quality assesment separated from the core pipeline.</p>

<p>REMOVE: No support for debian and RPM packages.</p>

<h3>SPAdes 2.0.1, 26 Apr 2012</h3>

<p>FIX: Quality tool fixed.</p>

<h3>SPAdes 2.0.0, 18 Apr 2012</h3>

<p>Initial release.</p>

<p>
    <a href="http://cab.spbu.ru/software/spades/">SPAdes website</a>
</p>

</body>
</html>

